wsmux/stream.go

```diff
@@ -0,0 +1,318 @@
+// reads can continue
+func (s *stream) PushAndBroadcast(buf []byte) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	defer s.session.logger.Printf("push broadcasted : stream %d", s.id)
+	_, err := s.b.Write(buf)
+	s.endErr = err
+}
+
+// AcceptStream accepts the current stream by closing the accepted channel
+func (s *stream) AcceptStream(read uint32) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	s.unblocked += read
+	s.state = accepted
+	close(s.accepted)
+
+}
+
+// SetDeadline sets the read and write deadlines for the stream
+func (s *stream) SetDeadline(t time.Time) error {
+	if err := s.SetReadDeadline(t); err != nil {
+		s.endErr = err
+		return err
+	}
+	if err := s.SetWriteDeadline(t); err != nil {
+		s.endErr = err
+		return err
+	}
+	return nil
+}
+
+func (s *stream) IsRemovable() bool {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.state == dead && s.b.Len() == 0
+}
+
+// setRemoteClosed sets the value of stream.remoteClosed. This indicates that the remote has sent a fin packet
+func (s *stream) setRemoteClosed() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	if s.state == closed {
+		s.state = dead
+	} else {
+		s.state = remoteClosed
+	}
+}
+
+//LocalAddr returns the local address of the underlying connection
+func (s *stream) LocalAddr() net.Addr {
+	return s.session.conn.LocalAddr()
+}
+
+// RemoteAddr returns the remote address of the underlying connection
+func (s *stream) RemoteAddr() net.Addr {
+	return s.session.conn.RemoteAddr()
+}
+
+// Close closes the stream
+func (s *stream) Close() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+
+	switch s.state {
+	// return nil if already closed
+	case closed:
+		return nil
+	case remoteClosed:
+		s.state = dead
+	default:
+		s.state = closed
+	}
+
+	if err := s.session.send(newFinFrame(s.id)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read reads bytes from the stream
+func (s *stream) Read(buf []byte) (int, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	s.session.logger.Printf("stream %d: read requested", s.id)
+
+	for s.b.Len() == 0 && s.endErr == nil {
+		// return EOF if remoteClosed (remoteClosed + deadOnEmpty + dead)
+		if s.state == remoteClosed || s.state == dead {
+			return 0, io.EOF
+		}
+
+		if s.readDeadlineExceeded {
+			return 0, ErrReadTimeout
+		}
+		s.session.logger.Printf("stream %d: read waiting", s.id)
+		s.c.Wait()
+	}
+
+	if s.endErr != nil {
+		return 0, s.endErr
+	}
+
+	n, _ := s.b.Read(buf)
+	if err := s.session.send(newAckFrame(s.id, uint32(n))); err != nil {
+		return n, err
+	}
+
+	s.session.logger.Printf("stream %d: read completed", s.id)
+
+	return n, nil
+}
+
+// Write writes bytes to the stream
+func (s *stream) Write(buf []byte) (int, error) {
+
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	w := 0
+	for len(buf) > 0 {
+		// if stream is closed or waiting to be empty then abort
+		if s.state == closed || s.state == dead {
```

Comment:
Same comments as on before... Try to use the loop only for waiting.... It's much simpler... There is a risk cases aren't handled correctly here...